storage: Fix and deflake TestRemoveRangeWithoutGC

bdarnell opened the pull request:
The fact that this test has been passing the vast majority of the time
has been due to false positives; the range was in fact getting GCed at
the start of the test but this asynchronous process was slow enough that
we didn't see it.

Fix the test by correctly disabling the replica GC queue and adding
additional checks to ensure that the data is still present after
restart.

Fixes #9506

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10540)
<!-- Reviewable:end -->

bdarnell committed:
storage: Fix and deflake TestRemoveRangeWithoutGC

The fact that this test has been passing the vast majority of the time
has been due to false positives; the range was in fact getting GCed at
the start of the test but this asynchronous process was slow enough that
we didn't see it.

Fix the test by correctly disabling the replica GC queue and adding
additional checks to ensure that the data is still present after
restart.

Fixes #9506

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :tada:

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2565 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-KW3eORAmOWJAdYaaO6o:-KW3eORAmOWJAdYaaO6p:b-e2d0jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/abcbbb66174a68eb34a63cd3a4eb0a690c07c30f/pkg/storage/client_raft_test.go#L2565)):*
> ```Go
> 		}
> 		desc := rep.Desc()
> 		if len(desc.Replicas) != 1 {
> ```

paranoid, but would feel slightly better about comparing this against the return of `rep.GetReplicaDescriptor()`

feel free to ignore since NYC

---

*[pkg/storage/client_raft_test.go, line 2595 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-KW3ezxxGw0KCUoNegWv:-KW3ezxxGw0KCUoNegWw:bnjds1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/abcbbb66174a68eb34a63cd3a4eb0a690c07c30f/pkg/storage/client_raft_test.go#L2595)):*
> ```Go
> 	// Re-enable the GC queue to allow the replica to be destroyed
> 	// (after the simulated passage of time).
> 	mtc.manualClock.Increment(int64(2*storage.ReplicaGCQueueInactivityThreshold + 1))
> ```

how come the `+ 1`?

---

*[pkg/storage/queue.go, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-KW3e7KRtn5dWSVs3ZSr:-KW3e7KRtn5dWSVs3ZSs:b-c2y88a) ([raw file](https://github.com/cockroachdb/cockroach/blob/abcbbb66174a68eb34a63cd3a4eb0a690c07c30f/pkg/storage/queue.go#L698)):*
> ```Go
> // Exposed for testing only.
> //
> // TODO(bdarnell): this method may race with the call to bq.pop() in
> ```

What's the TODO here? If you're not around, how do I or another contributor pick up this work?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-:-KW3f6BsS4tLSc1PkyCx:bck4n45)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 2595 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-KW3ezxxGw0KCUoNegWv:-KW72RQdqMBKu_5KcSru:b7hdeka) ([raw file](https://github.com/cockroachdb/cockroach/blob/abcbbb66174a68eb34a63cd3a4eb0a690c07c30f/pkg/storage/client_raft_test.go#L2595)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come the `+ 1`?
</details>
Cargo culted from other tests that use this value. (and the `2*` was from me hacking on something and I've replaced it with a call to expireLeases. Both are needed here).  I think there's some comparison using `>` when it should arguably be `>=`. 

---

*[pkg/storage/queue.go, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540#-KW3e7KRtn5dWSVs3ZSr:-KW739e_Kkb7JNSy7d-O:b-5c7yia) ([raw file](https://github.com/cockroachdb/cockroach/blob/abcbbb66174a68eb34a63cd3a4eb0a690c07c30f/pkg/storage/queue.go#L698)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What's the TODO here? If you're not around, how do I or another contributor pick up this work?
</details>
Added comment: This should synchronize
// somehow with processLoop so we wait for anything being handled
// there to finish too. When that's done, the SucceedsSoon at the end
// of TestRemoveRangeWithoutGC (and perhaps others) can be replaced
// with a one-time check.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10540)*
<!-- Sent from Reviewable.io -->
